### ISSUE TITLE
connect() doesn't return EISCONN when linked to socket creation

### ISSUE THREAD

**Issue by cmazakas** [axboe/liburing#980]
I noticed a really odd issue when writing tests for my io_uring runtime.

Namely, I'm using direct descriptors and I'm linking socket creation to a connect() I/O operation. The issue is, if I chain a normal non-linked connect() afterwards, the first run doesn't return `-EISCONN`. It actually returns `0`! Then subsequent connect() ops will properly return `-EISCONN`.

I'm curious if there's just an outright bug in my code.

Here's a minimal repro with the console output:
```
created acceptor socket
initialized the ring
{user_data: 2, res: 0}
{user_data: 3, res: 0}
{user_data: 1, res: 0}
{user_data: 3, res: 0} # this is the offending line, should be -106
{user_data: 3, res: -106}
```

uname output:
```bash
exbigboss@exbigboss-ubuntu:~/cpp/connect_bug$ uname -a
Linux exbigboss-ubuntu 6.5.0-10-generic #10-Ubuntu SMP PREEMPT_DYNAMIC Fri Oct 13 13:49:38 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

Build script:
```bash
clang++-17 \
  -std=c++20 -I ~/cpp/__install__/include \
  main.cpp ~/cpp/__install__/lib/liburing.a \
  -fsanitize=address,undefined \
  -o bug-test

./bug-test
```

`main.cpp`:
```cpp
#include <cassert>
#include <iostream>

#include <liburing.h>
#include <netinet/in.h>
#include <sys/socket.h>

int main() {
  int ret = -1;
  auto acceptor_fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
  assert(acceptor_fd >= 0);

  int const enable = 1;
  ret = setsockopt(acceptor_fd, SOL_SOCKET, SO_REUSEADDR, &enable,
                   sizeof(enable));

  if (ret == -1) {
    std::cerr << "failed setsockopt() with: " << errno << std::endl;
    return 1;
  }

  sockaddr_in server_addr = {};
  server_addr.sin_addr.s_addr = htonl(0x7f000001);
  server_addr.sin_family = AF_INET;
  server_addr.sin_port = htons(42069);

  ret = bind(acceptor_fd, reinterpret_cast<sockaddr const *>(&server_addr),
             sizeof(sockaddr_in));
  if (ret == -1) {
    std::cerr << "failed to bind the socket with: " << errno << std::endl;
    return 1;
  }

  ret = listen(acceptor_fd, 256);
  if (ret == -1) {
    std::cerr << "failed to set listen() state with: " << errno << std::endl;
    return 1;
  }

  std::cout << "created acceptor socket" << std::endl;

  io_uring ioring = {};
  auto ring = &ioring;
  ret = io_uring_queue_init(128, ring, 0);
  if (ret != 0) {
    std::cout << "failed to initialize the ring with: " << -ret << std::endl;
    return 1;
  }

  ret = io_uring_register_files_sparse(ring, 1024);
  if (ret < 0) {
    std::cout << "failed to register the sparse file table with: " << -ret
              << std::endl;
    return 1;
  }

  std::cout << "initialized the ring" << std::endl;

  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_accept_direct(sqe, acceptor_fd, nullptr, nullptr, 0, 42);
    io_uring_sqe_set_data64(sqe, 1);
  }

  io_uring_submit(ring);

  auto const file_index = 89;
  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_socket_direct(sqe, AF_INET, SOCK_STREAM, IPPROTO_TCP,
                                file_index, 0);
    io_uring_sqe_set_data64(sqe, 2);
    io_uring_sqe_set_flags(sqe, IOSQE_IO_LINK);
  }
  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_connect(sqe, file_index,
                          reinterpret_cast<sockaddr const *>(&server_addr),
                          sizeof(sockaddr_in));
    io_uring_sqe_set_data64(sqe, 3);
    io_uring_sqe_set_flags(sqe, IOSQE_IO_LINK | IOSQE_FIXED_FILE);
  }
  {
    __kernel_timespec ts = {.tv_sec = 3, .tv_nsec = 0};
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_link_timeout(sqe, &ts, 0);
    io_uring_sqe_set_data64(sqe, 4);
    io_uring_sqe_set_flags(sqe, IOSQE_CQE_SKIP_SUCCESS);
  }

  io_uring_submit_and_wait(ring, 3);

  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_connect(sqe, file_index,
                          reinterpret_cast<sockaddr const *>(&server_addr),
                          sizeof(sockaddr_in));
    io_uring_sqe_set_data64(sqe, 3);
    io_uring_sqe_set_flags(sqe, IOSQE_IO_LINK | IOSQE_FIXED_FILE);
  }
  {
    __kernel_timespec ts = {.tv_sec = 3, .tv_nsec = 0};
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_link_timeout(sqe, &ts, 0);
    io_uring_sqe_set_data64(sqe, 6);
    io_uring_sqe_set_flags(sqe, IOSQE_CQE_SKIP_SUCCESS);
  }

  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_connect(sqe, file_index,
                          reinterpret_cast<sockaddr const *>(&server_addr),
                          sizeof(sockaddr_in));
    io_uring_sqe_set_data64(sqe, 3);
    io_uring_sqe_set_flags(sqe, IOSQE_IO_LINK | IOSQE_FIXED_FILE);
  }
  {
    __kernel_timespec ts = {.tv_sec = 3, .tv_nsec = 0};
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_link_timeout(sqe, &ts, 0);
    io_uring_sqe_set_data64(sqe, 6);
    io_uring_sqe_set_flags(sqe, IOSQE_CQE_SKIP_SUCCESS);
  }

  io_uring_submit_and_wait(ring, 1);

  io_uring_cqe *cqes[5] = {};
  io_uring_peek_batch_cqe(ring, cqes, std::size(cqes));

  for (int i = 0; i < std::size(cqes); ++i) {
    std::cout << "{user_data: " << cqes[i]->user_data
              << ", res: " << cqes[i]->res << "}" << std::endl;
    ;
  }

  io_uring_queue_exit(ring);
  close(acceptor_fd);
}
```

**Comment by cmazakas** [comment]
I've done some more experimenting.

This issue seems to replicate even when not using direct descriptors and with non-linked requests.

It doesn't necessarily seem like io_uring is emplacing the CQE too soon because accept() finishes when it should but it seems like there's a strange lag in state that's independent of timing conditions too.

**Comment by cmazakas** [comment]
@axboe let me know what you think about this! I've removed the linked timeouts and other such things as the bug still replicates.

**Comment by axboe** [comment]
I was actually just taking a look at it this morning. io_uring just calls the lower level connect, and for the 2nd iteration on trying to connect, the socket is in a CONNECTING state which is why it returns 0. On the 3rd attempt, it's in CONNECTED state and you get ```-EISCONN``` as expected. So not quite sure what is going on here, didn't get much further just yet but I'm pondering if a normal connect(2) would yield the same.

**Comment by cmazakas** [comment]
lmao good timing!

I actually just tried using a blocking connect(2) call and it gave me the expected results.

Output:
```
created acceptor socket
initialized the ring
106
106
```

code:
```cpp
#include <cassert>
#include <iostream>

#include <liburing.h>
#include <netinet/in.h>
#include <sys/socket.h>

int main() {
  int ret = -1;
  auto acceptor_fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
  assert(acceptor_fd >= 0);

  int const enable = 1;
  ret = setsockopt(acceptor_fd, SOL_SOCKET, SO_REUSEADDR, &enable,
                   sizeof(enable));

  if (ret == -1) {
    std::cerr << "failed setsockopt() with: " << errno << std::endl;
    return 1;
  }

  sockaddr_in server_addr = {};
  server_addr.sin_addr.s_addr = htonl(0x7f000001);
  server_addr.sin_family = AF_INET;
  server_addr.sin_port = htons(42069);

  ret = bind(acceptor_fd, reinterpret_cast<sockaddr const *>(&server_addr),
             sizeof(sockaddr_in));
  if (ret == -1) {
    std::cerr << "failed to bind the socket with: " << errno << std::endl;
    return 1;
  }

  ret = listen(acceptor_fd, 256);
  if (ret == -1) {
    std::cerr << "failed to set listen() state with: " << errno << std::endl;
    return 1;
  }

  std::cout << "created acceptor socket" << std::endl;

  io_uring ioring = {};
  auto ring = &ioring;
  ret = io_uring_queue_init(128, ring, 0);
  if (ret != 0) {
    std::cout << "failed to initialize the ring with: " << -ret << std::endl;
    return 1;
  }

  ret = io_uring_register_files_sparse(ring, 1024);
  if (ret < 0) {
    std::cout << "failed to register the sparse file table with: " << -ret
              << std::endl;
    return 1;
  }

  std::cout << "initialized the ring" << std::endl;

  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_accept_direct(sqe, acceptor_fd, nullptr, nullptr, 0, 42);
    io_uring_sqe_set_data64(sqe, 1);
  }

  io_uring_submit(ring);

  auto sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
  ret = connect(sock, reinterpret_cast<sockaddr const *>(&server_addr),
                sizeof(server_addr));
  assert(ret >= 0);

  ret = connect(sock, reinterpret_cast<sockaddr const *>(&server_addr),
                sizeof(server_addr));
  if (ret < 0) {
    std::cerr << errno << std::endl;
  }

  ret = connect(sock, reinterpret_cast<sockaddr const *>(&server_addr),
                sizeof(server_addr));
  if (ret < 0) {
    std::cerr << errno << std::endl;
  }

  io_uring_queue_exit(ring);
  close(acceptor_fd);
}
```

**Comment by axboe** [comment]
Excellent! Then I don't have to test that. I'll dive into this and see what I find.

**Comment by axboe** [comment]
The difference is that your connect(2) test is using blocking sockets, so it waits for it. If you instantiate that socket() with SOCK_NONBLOCK, then you'd get:

```
created acceptor socket
initialized the ring
connect ret 115
connect ret 0
connect ret 106
```

which is basically what you see with io_uring - first one returns ```-EINPROGRESS``` and then get the proper connect, and then the third one will return ```-EISCONN```. Hmm...

**Comment by axboe** [comment]
You could try with this:

```
diff --git a/io_uring/net.c b/io_uring/net.c
index 7a8e298af81b..b45ac315ef90 100644
--- a/io_uring/net.c
+++ b/io_uring/net.c
@@ -1466,8 +1466,15 @@ int io_connect(struct io_kiocb *req, unsigned int issue_flags)
 
 		ret = -ENOTSOCK;
 		socket = sock_from_file(req->file);
-		if (socket)
+		if (socket) {
 			ret = sock_error(socket->sk);
+			if (!ret) {
+				/* no error, mark us connected */
+				lock_sock(socket->sk);
+				socket->state = SS_CONNECTED;
+				release_sock(socket->sk);
+			}
+		}
 		goto out;
 	}
 ```

**Comment by cmazakas** [comment]
dumb question, how should I apply this patch and rebuild io_uring?

**Comment by axboe** [comment]
Let's hang on a bit... I'm not really convinced this is a real issue. As mentioned, if you use a nonblocking socket and do:

1) connect(sock, ...); -> errno == ```-EINPROGRESS```
2) poll sock for POLLOUT, returns when connected
3) connect(sock, ...) -> ret == 0
4) connect(sock, ...) -> errno == ```-EISCONN```

is the same sequence you'd see with io_uring. IOW, it behaves like a non-blocking socket would. You could of course argue that step 3 should return ```-EISCONN``` as that would be expected, but the fact is that the networking stack doesn't make that guarantee.

Not sure this is worth doing much about, or at least if we do, we should take this up with the networking side. This has probably been the behavior for decades, maybe it's even documented or discussed somewhere already.

**Comment by axboe** [comment]
I think this is the saner fix:

```
diff --git a/io_uring/net.c b/io_uring/net.c
index eb1f51ddcb23..22e6a135e5e7 100644
--- a/io_uring/net.c
+++ b/io_uring/net.c
@@ -1456,16 +1456,6 @@ int io_connect(struct io_kiocb *req, unsigned int issue_flags)
 	int ret;
 	bool force_nonblock = issue_flags & IO_URING_F_NONBLOCK;
 
-	if (connect->in_progress) {
-		struct socket *socket;
-
-		ret = -ENOTSOCK;
-		socket = sock_from_file(req->file);
-		if (socket)
-			ret = sock_error(socket->sk);
-		goto out;
-	}
-
 	if (req_has_async_data(req)) {
 		io = req->async_data;
 	} else {
@@ -1485,9 +1475,7 @@ int io_connect(struct io_kiocb *req, unsigned int issue_flags)
 	    && force_nonblock) {
 		if (ret == -EINPROGRESS) {
 			connect->in_progress = true;
-			return -EAGAIN;
-		}
-		if (ret == -ECONNABORTED) {
+		} else if (ret == -ECONNABORTED) {
 			if (connect->seen_econnaborted)
 				goto out;
 			connect->seen_econnaborted = true;
@@ -1501,6 +1489,16 @@ int io_connect(struct io_kiocb *req, unsigned int issue_flags)
 		memcpy(req->async_data, &__io, sizeof(__io));
 		return -EAGAIN;
 	}
+	if (connect->in_progress) {
+		/*
+		 * At least bluetooth will return -EBADFD on a re-connect
+		 * attempt, and it's (supposedly) also valid to get -EISCONN
+		 * which means the previous result is good. For both of these,
+		 * grab the sock_error() and use that for the completion.
+		 */
+		if (ret == -EBADFD || ret == -EISCONN)
+			ret = sock_error(sock_from_file(req->file)->sk);
+	}
 	if (ret == -ERESTARTSYS)
 		ret = -EINTR;
 out:
```

as that should yield the correct result regardless. I'll test this now and send it out.

**Comment by cmazakas** [comment]
I see. You know, I figured a lot of this might also be because I'm misunderstanding TCP/networking.

I will say this, I definitely think this defies user expectation. It's as you note, this is how normal non-blocking sockets behave but I'm not using a non-blocking socket.

Once the CQE comes in, imo, all further connect() calls should result in a EISCONN.

That being said, I have a "workaround" that gives me the "correct" output:

```
created acceptor socket
initialized the ring
{user_data: 2, res: 0}
{user_data: 3, res: 0}
{user_data: 1, res: 0}
{user_data: 3, res: -106}
{user_data: 3, res: -106}
```

Basically, I link a connect() to a connect() lmao. It ain't pretty or clean but it does the job without incurring a huge perf penalty.

using:
```cpp
#include <cassert>
#include <iostream>

#include <liburing.h>
#include <netinet/in.h>
#include <sys/socket.h>

int main() {
  int ret = -1;
  auto acceptor_fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
  assert(acceptor_fd >= 0);

  int const enable = 1;
  ret = setsockopt(acceptor_fd, SOL_SOCKET, SO_REUSEADDR, &enable,
                   sizeof(enable));

  if (ret == -1) {
    std::cerr << "failed setsockopt() with: " << errno << std::endl;
    return 1;
  }

  sockaddr_in server_addr = {};
  server_addr.sin_addr.s_addr = htonl(0x7f000001);
  server_addr.sin_family = AF_INET;
  server_addr.sin_port = htons(42069);

  ret = bind(acceptor_fd, reinterpret_cast<sockaddr const *>(&server_addr),
             sizeof(sockaddr_in));
  if (ret == -1) {
    std::cerr << "failed to bind the socket with: " << errno << std::endl;
    return 1;
  }

  ret = listen(acceptor_fd, 256);
  if (ret == -1) {
    std::cerr << "failed to set listen() state with: " << errno << std::endl;
    return 1;
  }

  std::cout << "created acceptor socket" << std::endl;

  io_uring ioring = {};
  auto ring = &ioring;
  ret = io_uring_queue_init(128, ring, 0);
  if (ret != 0) {
    std::cout << "failed to initialize the ring with: " << -ret << std::endl;
    return 1;
  }

  ret = io_uring_register_files_sparse(ring, 1024);
  if (ret < 0) {
    std::cout << "failed to register the sparse file table with: " << -ret
              << std::endl;
    return 1;
  }

  std::cout << "initialized the ring" << std::endl;

  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_accept_direct(sqe, acceptor_fd, nullptr, nullptr, 0, 42);
    io_uring_sqe_set_data64(sqe, 1);
  }

  io_uring_submit(ring);

  auto const file_index = 89;
  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_socket_direct(sqe, AF_INET, SOCK_STREAM, IPPROTO_TCP,
                                file_index, 0);
    io_uring_sqe_set_data64(sqe, 2);
    io_uring_sqe_set_flags(sqe, IOSQE_IO_LINK);
  }
  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_connect(sqe, file_index,
                          reinterpret_cast<sockaddr const *>(&server_addr),
                          sizeof(sockaddr_in));
    io_uring_sqe_set_data64(sqe, 3);
    io_uring_sqe_set_flags(sqe, IOSQE_IO_LINK | IOSQE_FIXED_FILE);
  }
  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_connect(sqe, file_index,
                          reinterpret_cast<sockaddr const *>(&server_addr),
                          sizeof(sockaddr_in));
    io_uring_sqe_set_data64(sqe, 3);
    io_uring_sqe_set_flags(sqe, IOSQE_IO_LINK | IOSQE_FIXED_FILE |
                                    IOSQE_CQE_SKIP_SUCCESS);
  }
  {
    __kernel_timespec ts = {.tv_sec = 3, .tv_nsec = 0};
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_link_timeout(sqe, &ts, 0);
    io_uring_sqe_set_data64(sqe, 4);
    io_uring_sqe_set_flags(sqe, IOSQE_CQE_SKIP_SUCCESS);
  }

  io_uring_submit_and_wait(ring, 3);

  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_connect(sqe, file_index,
                          reinterpret_cast<sockaddr const *>(&server_addr),
                          sizeof(sockaddr_in));
    io_uring_sqe_set_data64(sqe, 3);
    io_uring_sqe_set_flags(sqe, IOSQE_FIXED_FILE);
  }

  {
    auto sqe = io_uring_get_sqe(ring);
    io_uring_prep_connect(sqe, file_index,
                          reinterpret_cast<sockaddr const *>(&server_addr),
                          sizeof(sockaddr_in));
    io_uring_sqe_set_data64(sqe, 3);
    io_uring_sqe_set_flags(sqe, IOSQE_FIXED_FILE);
  }

  io_uring_submit_and_wait(ring, 1);

  io_uring_cqe *cqes[5] = {};
  io_uring_peek_batch_cqe(ring, cqes, std::size(cqes));

  for (int i = 0; i < std::size(cqes); ++i) {
    std::cout << "{user_data: " << cqes[i]->user_data
              << ", res: " << cqes[i]->res << "}" << std::endl;
    ;
  }

  io_uring_queue_exit(ring);
  close(acceptor_fd);
}
```

**Comment by axboe** [comment]
I posted a fix here:

https://lore.kernel.org/io-uring/893c142c-8673-4a8a-9496-c2f13e7ed746@kernel.dk/

which should do the right thing for both the original issue that actually caused this change, and handle your case as well. Before that referenced bluetooth issue, io_uring did do the double connect itself. With the above, we do that again, but just ensure we handle potential issues that can arise from that. Eg your above comment will fail on bluetooth. Probably not an issue for you, but it is for me as it needs to work everywhere.

**Comment by axboe** [comment]
I'll get this fix into 6.7-rc1 and ensure it rolls back to stable as well.

**Comment by cmazakas** [comment]
Alright, that sounds perfect.

I really appreciate you taking the time to look into this.

I'll go ahead and close the issue. Users have a workaround for now and it sounds like the issue will eventually resolve itself. There's nothing else to be done. Thank you for your time.
